### PR TITLE
Xin testing blog PR

### PR DESCRIPTION
**Pull request by danscarborough**
Please add a direct link to your post here:

https://xchenscottlogic.github.io/blog/2023/10/29/testing-LLM-based-applications-strategy-and-challenges.html

Have you (please tick each box to show completion):

 - [x] Added your blog post to a single category?
 - [x] Added a brief summary for your post? Summaries should be roughly two sentences in length and give potential readers a good idea of the contents of your post.
 - [x] Checked that the build passes?
 - [x] Checked your spelling (you can use `npm spellcheck` if that's your thing)
 - [x] Ensured that your author profile contains a profile image, and a brief description of yourself? (make it more interesting than just your job title!)
 - [x] Optimised any images in your post? They should be less than 100KBytes as a general guide.

Posts are approved based on their category. replace `(at)` with `@` to notify them!

 - Tech - Colin Eberhardt - @ColinEberhardt
 - Cloud - Colin Eberhardt - (at)ColinEberhardt
 - Delivery - Matt Phillips - (at)scottlogicmphillips
 - UX - Graham Odds - (at)godds
 - People - Graham Odds - (at)godds
 - Testing - Laurence Pisani - @lpisani-SL

### CODE REVIEW

File: _posts/2023-10-29-testing-LLM-based-applications-strategy-and-challenges.markdown [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+The quality of LLM-based applications heavily relies on the quality of the model. A higher-performing model will undoubtedly produce more satisfying results. There are essentially two types of models. The first type is proprietary, such as OpenAI's GPT-3.5 and GPT-4, which appears as a black box to users but excels in handling complex tasks with higher performance. The second type is open source models, which may permit you to train them with your own specific data in addition to the language model. Model version, and prompt parameters, such as temperature, will also be essential in influencing the eventual output.
+
+The following testing metrics provide comprehensive measurements of the quality of LLM-based applications.
+
+![Test_Scottbot_0.png]({{ site.github.url }}/uploads/Test_Scottbot_0.png)
+
+
+## **Testing Strategy on LLM-Based Applications**
+
+Researchers commonly utilise standard metrics to assess traditional language models. These metrics include Perplexity (PPL), BLEU, and ROUGE. However, these metrics are not typically applicable for evaluating applications in production that utilise an LLM as a third-party service.
```
Comment (ColinEberhardt) [inline _posts/2023-10-29-testing-LLM-based-applications-strategy-and-challenges.markdown]: >  Perplexity (PPL), BLEU, and ROUGE

Could you provide links to these?

Comment (ColinEberhardt) [inline _posts/2023-10-29-testing-LLM-based-applications-strategy-and-challenges.markdown]: > these metrics are not typically applicable for evaluating applications in production that utilise an LLM as a third-party service.

that's an interesting point - why is that?

File: _posts/2023-10-29-testing-LLM-based-applications-strategy-and-challenges.markdown [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+
+To assess the correctness of LLM-based applications, the test cases should encompass fundamental use cases tailored to the application, while also taking into account potential user behaviours. In simpler terms, test cases should reflect what users intend to accomplish. Customised applications often include embedding domain-specific terminologies and knowledge, within an existing baseline LLM. The test cases should primarily address domain-specific scenarios.
+
+### **Use Case Study: Testing Scottbot**
+
+In this section, I will present the testing methodologies by using our customised chatbot, Scottbot, as an example. Scottbot is a Scott Logic domain-specific chatbot powered by GPT technology. It has embedded access to Scott Logic’s internal Confluence pages. In addition, it utilises Google and Wikipedia to respond to more general queries.
```
Comment (ColinEberhardt) [inline _posts/2023-10-29-testing-LLM-based-applications-strategy-and-challenges.markdown]: >  Scott Logic’s internal Confluence

Probably worth expanding on this - it has access to our intranet that is our internal knowledge-base

File: _posts/2023-10-29-testing-LLM-based-applications-strategy-and-challenges.markdown [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+
+In this section, I will present the testing methodologies by using our customised chatbot, Scottbot, as an example. Scottbot is a Scott Logic domain-specific chatbot powered by GPT technology. It has embedded access to Scott Logic’s internal Confluence pages. In addition, it utilises Google and Wikipedia to respond to more general queries. 
+
+The main objective of testing Scottbot is to guarantee accurate and meaningful responses to user queries. This involves system testing at various levels, with a focus on end-to-end scenarios. Functional testing is carried out through the pytest automation framework, supplemented by manual verification. Security testing primarily depends on manual verification to ensure Scottbot's robustness and compliance with legal and ethical requirements. Performance testing is currently not executed but remains a consideration for future assessments.
+
+As we use pytest as the automation framework, pytest-bdd is a natural choice to give us a BDD style for all of our scenarios. As Scottbot is deployed on Azure, automated tests can be executed through its CI/CD pipeline. Azure also provides a dashboard for monitoring and reporting the test results.
```
Comment (ColinEberhardt) [inline _posts/2023-10-29-testing-LLM-based-applications-strategy-and-challenges.markdown]: > pytest-bdd

link please!

File: _posts/2023-10-29-testing-LLM-based-applications-strategy-and-challenges.markdown [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+
+Based on Generative AI's non-deterministic feature, we can't do the exact match for the test results. A good solution to solve this issue is to introduce the LangChain string evaluators which uses a specified language model to evaluate a predicted string for a given input. We use LLMs via this evaluator framework to evaluate both the correctness and the confidence of the responses in our tests.
```
Comment (ColinEberhardt) [inline _posts/2023-10-29-testing-LLM-based-applications-strategy-and-challenges.markdown]: > Based on Generative AI's non-deterministic feature

It's probably worth expanding on this - I'm not sure it is common knowledge that reducing temperature to zero still resuts in variability in the output

File: _posts/2023-10-29-testing-LLM-based-applications-strategy-and-challenges.markdown [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+
+As we use pytest as the automation framework, pytest-bdd is a natural choice to give us a BDD style for all of our scenarios. As Scottbot is deployed on Azure, automated tests can be executed through its CI/CD pipeline. Azure also provides a dashboard for monitoring and reporting the test results.
+
+Based on Generative AI's non-deterministic feature, we can't do the exact match for the test results. A good solution to solve this issue is to introduce the LangChain string evaluators which uses a specified language model to evaluate a predicted string for a given input. We use LLMs via this evaluator framework to evaluate both the correctness and the confidence of the responses in our tests. 
+
+#### **1. Verify the Factual Correctness of the Responses**
+It ensures that the information provided by LLM applications aligns with real-world facts. Hallucination is a weakness of LLM. As a result, accuracy is crucial, particularly in customised applications, as information must be sourced from a reliable domain knowledge base to reduce the impact of hallucination.
```
Comment (ColinEberhardt) [inline _posts/2023-10-29-testing-LLM-based-applications-strategy-and-challenges.markdown]: Is it worth pointing out here that using the correct source is quite critical for Scottbot? We need to ensure that it chooses between Confluence and the internet wisely!

File: _posts/2023-10-29-testing-LLM-based-applications-strategy-and-challenges.markdown [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+Test cases: 
+
+<script src="https://gist.github.com/XChenscottlogic/68e297d1aeb588d3acb86ba878e0e495.js"></script>
+
+#### **5. Verify the Readability of the Responses**
+Readability in LLM-applications ensures that the provided responses are logically and linguistically coherent, easily understood, and conform to the expected format. Temperature serves as a parameter to determine the creativity level of LLM-based applications. Increasing the temperature results in more diverse and creative responses. As a consequence, when the temperature setting is elevated, there is a potential impact on response readability. In the most extreme situation, the output might consist of arbitrary and nonsensical characters.
+
+Test structure:
+
+![Test_Scottbot_3_2.png]({{ site.github.url }}/uploads/Test_Scottbot_3_2.png)
```
Comment (ColinEberhardt) [inline _posts/2023-10-29-testing-LLM-based-applications-strategy-and-challenges.markdown]: How do you assert that the output "should be readable"

**Review by ColinEberhardt — CHANGES_REQUESTED** [review]
Thanks - a really interesting post. Added a few specific comments, also a couple of other points ...

Some of your example BDD files are quite long, perhaps edit them down to smaller set of examples?

For some of the BDD tests I am unsure how they are actually implemented, for example, they verify that the output "should be readable", the "response time should be reasonable", the response "should be appropriate"

File: _posts/2023-10-29-testing-LLM-based-applications-strategy-and-challenges.markdown (hunk 6 above)
Comment (XChenscottlogic) [reply _posts/2023-10-29-testing-LLM-based-applications-strategy-and-challenges.markdown]: For the time being, it is a human verification process to check the readability. We may want to check if the contents contain any rubbish code, like consequent wild card characters in an automation way.

File: _posts/2023-10-29-testing-LLM-based-applications-strategy-and-challenges.markdown (hunk 5 above)
Comment (XChenscottlogic) [reply _posts/2023-10-29-testing-LLM-based-applications-strategy-and-challenges.markdown]: Updated.

File: _posts/2023-10-29-testing-LLM-based-applications-strategy-and-challenges.markdown (hunk 3 above)
Comment (XChenscottlogic) [reply _posts/2023-10-29-testing-LLM-based-applications-strategy-and-challenges.markdown]: Updated.

File: _posts/2023-10-29-testing-LLM-based-applications-strategy-and-challenges.markdown (hunk 2 above)
Comment (XChenscottlogic) [reply _posts/2023-10-29-testing-LLM-based-applications-strategy-and-challenges.markdown]: Updated.

File: _posts/2023-10-29-testing-LLM-based-applications-strategy-and-challenges.markdown (hunk 1 above)
Comment (XChenscottlogic) [reply _posts/2023-10-29-testing-LLM-based-applications-strategy-and-challenges.markdown]: Updated.

Comment (XChenscottlogic) [reply _posts/2023-10-29-testing-LLM-based-applications-strategy-and-challenges.markdown]: Updated.

File: _posts/2023-10-29-testing-LLM-based-applications-strategy-and-challenges.markdown (hunk 4 above)
Comment (XChenscottlogic) [reply _posts/2023-10-29-testing-LLM-based-applications-strategy-and-challenges.markdown]: We set temperature to zero but my observation is that each time Scottbot still gives various responses, based on the question it received.  I looked on the internet and it is said zero temperature means completely deterministic. I am confused.

Comment (XChenscottlogic) [reply _posts/2023-10-29-testing-LLM-based-applications-strategy-and-challenges.markdown]: Updated.
